(8.0) PXC-3821: Truncating performance_schema's table in one node, replicating the truncate statement in other nodes also

kamil-holubicki opened the pull request:
https://jira.percona.com/browse/PXC-3821

Problem:
Truncation of P_S table is replicated across the cluster

Cause:
Truncation of P_S tables is done exactly the same as for any other table, so the query is replicated as TOI.

Solution:

Filter out P_S tables truncation and do not initiate TOI for them.
For the case when there is source node without this fix, filter out truncation on the replica node in the same way as it was introduced in commit https://github.com/percona/percona-xtradb-cluster/commit/44c56d37dfe767c31b5ca80812229881c32140f3 (is_executed_by_slave() function fixed to be aware of wsrep applier thread).

kamil-holubicki committed:
PXC-3821: Truncating performance_schema's table in
one node, replicating the truncate statement in
other nodes also.

https://jira.percona.com/browse/PXC-3821

Problem:
Truncation of P_S table is replicated across the cluster

Cause:
Truncation of P_S tables is done exactly the same as for any other
table, so the query is replicated as TOI.

Solution:
1. Filter out P_S tables truncation and do not initiate TOI for them.
2. For the case when there is source node without this fix, filter out
truncation on the replica node in the same way as it was introduced in
commit 44c56d37dfe767c31b5ca80812229881c32140f3 (is_executed_by_slave()
function fixed to be aware of wsrep applier thread).

kamil-holubicki commented:
https://pxc.cd.percona.com/job/pxc-8.0-pipeline-parallel-mtr/751/

venkatesh-prasad-v commented on mysql-test/suite/galera/t/galera_truncate.test:
```diff
@@ -103,3 +103,34 @@ TRUNCATE TABLE t1;
+--disable_result_log
+
+--connection node_2
+# We need to execute SELECT COUNT before geting the actual result to have it already in events_statements_summary_by_digest table 
```
```suggestion
# We need to execute SELECT COUNT before getting the actual result to have it already in events_statements_summary_by_digest table 
```

kamil-holubicki replied:
Fixed in 5.7 PR. 
Once https://github.com/percona/percona-xtradb-cluster/pull/1697 is approved, I will merge it to this PR according to GCA.

venkatesh-prasad-v commented on mysql-test/suite/galera/t/galera_truncate.test:
```diff
@@ -103,3 +103,34 @@ TRUNCATE TABLE t1;
+--connection node_1
+TRUNCATE TABLE performance_schema.events_statements_summary_by_digest;
+
+# It should be 1 at the moment. We will not use assert.inc directly, as it sends more queries, and the result is not deterministic.
+--let $node_1_count = `SELECT COUNT(*) FROM performance_schema.events_statements_summary_by_digest`
+--let $assert_text = "performance_schema.events_statements_summary_by_digest table should be truncated"
+--let $assert_cond = $node_1_count = 1
+--source include/assert.inc
```
Here we say that 
> We will not use assert.inc directly

But we are using assert.inc. Is this expected?

kamil-holubicki replied:
Yes, this is fine. The point is to not use it like
`--let $assert_cond = COUNT(*) = 1 FROM performance_schema.events_statements_summary_by_digest`

venkatesh-prasad-v replied:
Ok. Got it. Thanks for the clarification.

kamil-holubicki committed:
PXC-3821: Truncating performance_schema's table in
one node, replicating the truncate statement in
other nodes also.

https://jira.percona.com/browse/PXC-3821

Problem:
Truncation of P_S table is replicated across the cluster

Cause:
Truncation of P_S tables is done exactly the same as for any other
table, so the query is replicated as TOI.

Solution:
1. Filter out P_S tables truncation and do not initiate TOI for them.
2. For the case when there is source node without this fix, filter out
truncation on the replica node in the same way as it was introduced in
commit 44c56d37dfe767c31b5ca80812229881c32140f3 (is_executed_by_slave()
function fixed to be aware of wsrep applier thread).

kamil-holubicki commented:
https://ps80.cd.percona.com/view/Parallel%20MTR/job/percona-server-8.0-pipeline-parallel-mtr/289/
